mm0-rs/src/elab/lisp/debug.rs

```diff
@@ -0,0 +1,1465 @@
+        // Calling atom_id.dbg_aux(..) recursively here will loop and blow the stack.
+        match sort.atom {
+          AtomID(id) => ron.add_field("atom_id", id.to_string())
+        }
+        ron.add_field("name", sort.name.to_string());
+        ron.add_field("span", sort.span.dbg_aux(fe, p));
+        ron.add_field("full", sort.full.dbg_aux(fe, p));
+        ron.add_field("mods", sort.mods.to_string());
+        ron.to_doc(p)
+      }
+    }
+  }
+}
+
+impl EnvDebug for Sort {
+  fn dbg_aux<'a>(
+    &self, 
+    fe : FormatEnv<'a>,
+    p : &mut Printer<'a>, 
+  ) -> DocPtr<'a> {
+      let mut ron = RonStruct::new();
+      ron.add_name("Sort");
+      match self.atom {
+        AtomID(id) => ron.add_field("atom_id", id.to_string())
+      }
+      ron.add_field("name", self.name.to_string());
+      ron.add_field("span", self.span.dbg_aux(fe, p));
+      ron.add_field("full", self.full.dbg_aux(fe, p));
+      ron.add_field("mods", self.mods.to_string());
+      ron.to_doc(p)
+  }
+}
+
+impl EnvDebug for Span {
+  fn dbg_aux<'a>(
+    &self, 
+    _ : FormatEnv<'a>,
+    p : &mut Printer<'a>, 
+  ) -> DocPtr<'a> {
+    format!("{:#?}", self).alloc(p)
+  }
+}
+
+impl EnvDebug for FileSpan {
+  fn dbg_aux<'a>(
+    &self, 
+    fe : FormatEnv<'a>,
+    p : &mut Printer<'a>, 
+  ) -> DocPtr<'a> {
+    let mut ron = RonStruct::new();
+    ron.add_name("FileSpan");
+    ron.add_field("file", self.file.dbg_aux(fe, p));
+    ron.add_field("span", self.span.dbg_aux(fe, p));
+    ron.to_doc(p)
+  }
+}
+
+impl EnvDebug for FileRef {
+  fn dbg_aux<'a>(
+    &self, 
+    _ : FormatEnv<'a>,
+    p : &mut Printer<'a>, 
+  ) -> DocPtr<'a> {
+    let mut ron = RonStruct::new();
+    ron.add_name("FileRef");
+    ron.add_field("url", self.url().to_string());
+    ron.add_field("path", self.path().display().to_string());
+    ron.add_field("rel", self.rel().to_string());
+    ron.to_doc(p)
+  }
+}
+
+impl EnvDebug for Type {
+  fn dbg_aux<'a>(
+    &self, 
+    fe : FormatEnv<'a>,
+    p : &mut Printer<'a>, 
+  ) -> DocPtr<'a> {
+    let mut ron = RonTuple::new();
+    ron.add_name("Type");
+    match self {
+      Type::Bound(sort_id) => {
+        ron.add_name("Bound");
+        ron.add_field(sort_id.dbg_aux(fe, p));
+
+      },
+      Type::Reg(sort_id, n) => {
+        ron.add_name("Reg");
+        ron.add_field(sort_id.dbg_aux(fe, p));
+        ron.add_field(n.to_string());
+      }
+    }
+    ron.to_doc(p)
+  }
+}
+
+impl EnvDebug for Expr {
+  fn dbg_aux<'a>(
+    &self, 
+    fe : FormatEnv<'a>,
+    p : &mut Printer<'a>, 
+  ) -> DocPtr<'a> {
+    let mut ron = RonStruct::new();
+    ron.add_name("Expr");
+    ron.add_field("heap", self.heap.dbg_aux(fe, p));
+    ron.add_field("head", self.head.dbg_aux(fe, p));
+    ron.to_doc(p)
+  }
+}
+
+impl EnvDebug for TermID {
+  fn dbg_aux<'a>(
+    &self, 
+    fe : FormatEnv<'a>,
+    p : &mut Printer<'a>, 
+  ) -> DocPtr<'a> {
+    let term = &fe.terms[*self];
+    let mut ron = RonStruct::new();
```

Comment:
Same deal here. A `TermID` is not a `Term`, and if you try printing one as such (especially if you print the full proof of every term on every reference to it) you will get exponential blowup. Printing a `TermID` should just print the name of the term, while printing a `Term` should give the full story.